From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Wed, 15 Feb 2023 01:13:24 +0500
Subject: [PATCH] feat: drop use packets while inside the inventory

Will help to get rid of the desynchronization of the number of numPlayersUsing in chests/shulkers.
However, it modifies vanilla logic and could lead to problems with some plugins.

diff --git a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
index e182114ec0c5024485dcc284ea4bc927af62f449..c3dc3cb84dee84057189eee9e29a4452fb2edddd 100644
--- a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
+++ b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
@@ -290,6 +290,11 @@ public class ReaperConfig {
         mapDecorationsLimit = getInt("map-decorations-limit", 10);
     }
 
+    public static boolean dropUsePacketsInInventory;
+    private static void dropUsePacketsInInventory() {
+        dropUsePacketsInInventory = getBoolean("drop-use-packets-in-inventory", false);
+    }
+
     // Paper start - packet limiter
     public static String packetLimiterKickMessage;
     public static PacketLimit packetLimiterAllPackets;
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 063b6d0d05ca6bd47eab8e3e4268945dec89a492..329eb0dec15845e8f8f2df2afadd5df473929631 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1009,6 +1009,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
         PlayerConnectionUtils.ensureMainThread(packetplayinuseitem, this, this.player.x());
         if (this.player.isFrozen()) return; // CraftBukkit
         if (!checkLimit(packetplayinuseitem.timestamp)) return; // Spigot - check limit
+        if (player.activeContainer != player.defaultContainer && com.github.ruviolence.reaper.ReaperConfig.dropUsePacketsInInventory) return; // Reaper
         WorldServer worldserver = this.minecraftServer.getWorldServer(this.player.dimension);
         EnumHand enumhand = packetplayinuseitem.c();
         ItemStack itemstack = this.player.b(enumhand);
